src/main/java/org/spongepowered/api/text/dictionary/Dictionary.java

```diff
@@ -0,0 +1,394 @@
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * Utility class to allow plugin developers to easily add multi language support
+ * to their plugins.
+ */
+public class Dictionary implements Function<String, Function<Locale, Text>> {
+
+    private static final BiFunction<Locale, Map<Locale, Text>, Text> TRANSLATION_FALLBACK = (locale, map) -> map.values().iterator().next();
+    private final Map<String, Map<Locale, Text>> messages = new HashMap<>();
+    private final Function<String, String> messageFallback;
+    private final BiFunction<Locale, Map<Locale, Text>, Text> translationFallback;
+
+    /**
+     * Creates a new dictionary with no message fallback and a first result
+     * translation fallback.
+     */
+    public Dictionary() {
+        this(Functions.constantNull(), BiFunctions.constantNull());
+    }
+
+    /**
+     * Creates a new dictionary with the given message and translation fallback.
+     *
+     * @param messageFallback The message fallback function to use if the
+     *        requested translaion id could not be found
+     * @param translationFallback The translation fallback to use if the
+     *        requested there is no translation for the given language, but
+     *        translations for other languages
+     */
+    public Dictionary(Function<String, String> messageFallback,
+            BiFunction<Locale, Map<Locale, Text>, Text> translationFallback) {
+        super();
+        this.messageFallback = checkNotNull(messageFallback, "messageFallback");
+        this.translationFallback = BiFunctions.nonNullResultOrElse(
+                checkNotNull(translationFallback, "translationFallback"), TRANSLATION_FALLBACK);
+    }
+
+    /**
+     * Adds the given translation to this dictionaries storage.
+     *
+     * @param translationId The translation id to register the translation for
+     * @param locale The locale to register the translation for
+     * @param translation The translation to register
+     */
+    public void addTranslation(String translationId, Locale locale, Text translation) {
+        Map<Locale, Text> translations = this.messages.get(translationId);
+        if (translations == null) {
+            translations = new HashMap<>();
+            this.messages.put(translationId, translations);
+        }
+        translations.put(locale, translation);
+    }
+
+    /**
+     * Helper method to get a map containing a {@link Text} that the requested
+     * message could not be found
+     *
+     * @param translationId The translation id that could not be found
+     * @return A map containing the message not found text
+     */
+    protected Map<Locale, Text> getMissingTranslation(String translationId) {
```

Comment:
I don't agree with wrapping the key into this message, means that you have to ensure that all your keys are valid before trying to translate them or you get this super ugly message in the middle of your whatever. If a key can't be translated then you should simply return the key (this is the vanilla behavior).